utils/swift_build_support/swift_build_support/products/tsan_libdispatch.py

```diff
@@ -0,0 +1,58 @@
+# swift_build_support/products/tsan_libdispatch.py --------------*- python -*-
+#
+# This source file is part of the Swift.org open source project
+#
+# Copyright (c) 2014 - 2019 Apple Inc. and the Swift project authors
+# Licensed under Apache License v2.0 with Runtime Library Exception
+#
+# See https://swift.org/LICENSE.txt for license information
+# See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+#
+# ----------------------------------------------------------------------------
+
+import os
+
+from . import product
+from .. import shell
+
+
+class TSanLibDispatch(product.Product):
+    @classmethod
+    def product_source_name(cls):
+        return "tsan-libdispatch-test"
+
+    @classmethod
+    def is_build_script_impl_product(cls):
+        return False
+
+    def build(self, host_target):
+        """Build TSan runtime (compiler-rt)."""
+        rt_source_dir = os.path.join(self.source_dir, '../compiler-rt')
+        libdispatch_path = os.path.join(self.args.install_destdir, 'usr')
+        llvm_build_dir = os.path.join(self.build_dir, '../llvm-' + host_target)
+        cc_path = os.path.join(llvm_build_dir, 'bin/clang')
```

Comment:
What would happen if instead of using llvm_build_dir, you used the provided toolchain? Would this work? We really want all things in this compile step to use the toolchain. Are you linking against LLVM libraries?

Reply:
I tried it, but I couldn't make it work. I *think* since compiler-rt tests are not self-contained it needs stuff like cmake build rules and llvm-lit.

Reply:
How did it fail? We can change the "default toolchain install" to include llvm-lit.

Reply:
```
cmake -GNinja -B/home/yln/work/swift-source/build/buildbot_incremental_tsan_libdispatch_test/tsanlibdispatch-linux-x86_64 -DCMAKE_PREFIX_PATH=/home/yln/work/swift-source/build/buildbot_incremental_tsan_libdispatch_test/toolchain-linux-x86_64/usr -DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang++ -DCMAKE_BUILD_TYPE=Release -DLLVM_ENABLE_ASSERTIONS=ON -DCOMPILER_RT_INCLUDE_TESTS=ON -DCOMPILER_RT_INTERCEPT_LIBDISPATCH=ON -DCOMPILER_RT_LIBDISPATCH_INSTALL_PATH=/home/yln/work/swift-source/build/buildbot_incremental_tsan_libdispatch_test/toolchain-linux-x86_64/usr /home/yln/work/swift-source/compiler-rt

CMake Warning at cmake/Modules/CompilerRTUtils.cmake:213 (message):
  UNSUPPORTED COMPILER-RT CONFIGURATION DETECTED: llvm-config not found.

  Reconfigure with -DLLVM_CONFIG_PATH=path/to/llvm-config.
Call Stack (most recent call first):
  CMakeLists.txt:69 (load_llvm_config)

CMake Error at cmake/Modules/CompilerRTUtils.cmake:317 (string):
  string sub-command REPLACE requires at least four arguments.
Call Stack (most recent call first):
  CMakeLists.txt:149 (construct_compiler_rt_default_triple)

-- Compiler-RT supported architectures: 
-- Builtin supported architectures: 
CMake Error at unittests/CMakeLists.txt:2 (include):
  include could not find load file:

    AddLLVM

CMake Error at cmake/Modules/AddCompilerRT.cmake:663 (set_llvm_build_mode):
  Unknown CMake command "set_llvm_build_mode".
Call Stack (most recent call first):
  unittests/CMakeLists.txt:4 (configure_compiler_rt_lit_site_cfg)
```

Reply:
AddLLVM seems to refer to to llvm/cmake/modules/AddLLVM.cmake, i.e., even for standalone compiler_rt builds we still need some of the Cmake LLVM infrastructure. I don't think those things should end up in the toolchain.